[stable/redis-ha] Made the shutdown process more cluster aware

jasquat opened the pull request:
* added start scripts for the redis and sentinel containers that contain traps for proper shutdown
* Followed process recommended by https://redis.io/topics/sentinel under "Adding or removing Sentinels" heading
* increased termination grace period to 240 seconds since we do not want our graceful shutdown process interrupted
* made init script recover better when it first picks a master that does not exist rather than immediately forcing a failover
* updated README to mention acceptance tests that we used to improve scripts and verify cluster recovers

FYI: @ssalaues

Commit:
Improved probes, ha/stability fixes, and minor fixes

Signed-off-by: Salim <salim.salaues@scality.com>

Commit:
Remove skip defaults

Signed-off-by: Salim <salim.salaues@scality.com>

Commit:
fix os distinction

Signed-off-by: Salim <salim.salaues@scality.com>

Commit:
Fix: delete unused env var

Signed-off-by: Salim <salim.salaues@scality.com>

jasquat committed:
[stable/redis-ha] Made the shutdown process more cluster aware:

* added start scripts for the redis and sentinel containers that
contain traps for proper shutdown
* Followed process recommended by https://redis.io/topics/sentinel
under "Adding or removing Sentinels" heading
* increased termination grace period to 240 seconds since we do not
want our graceful shutdown process interrupted
* made init script recover better when it first picks a master that
does not exist rather than immediately forcing a failover

Co-authored-by: burnettk

Signed-off-by: Jason Lantz <jasquat@gmail.com>

jasquat committed:
[stable/redis-ha] bumped chart version to 3.0.4

Signed-off-by: Jason Lantz <jasquat@gmail.com>

jasquat committed:
[stable/redis-ha] updated README to mention tests

Signed-off-by: Jason Lantz <jasquat@gmail.com>

ssalaues commented on stable/redis-ha/templates/redis-ha-configmap.yaml:
```diff
@@ -36,12 +36,146 @@ data:
+        wait_for_redis_container_to_shutdown
+      fi
+    }
+    graceful_shutdown() {
```
The graceful shutdowns could be ran as part of the `preStop` container hook. This would allow for simplification of the scripts. 
https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/#container-hooks
https://pracucci.com/graceful-shutdown-of-kubernetes-pods.html

Something like:
1. pod gets marked for deletion, triggering the `preStop` hook
2. runs graceful shutdown script as `preStop` hook
3. script forces failover and waits for new master
4. script runs `redis-cli -h localhost SHUTDOWN` to gracefully terminate
5. once the preHook is finished the TERM signal gets sent by k8s to remove the pod

This way processes don't need to be forked, no need to track PIDs, and the start commands don't need to be wrapped shell scripts that sleep/wait.

jasquat replied:
Thanks for taking a look @ssalaues.

@burnettk and I merged your newest changes from master into our branch and attempted using the preStop hooks. It doesn't seem like this will work however. The shutdown process is:

In Sentinel Container:

1. ask redis if current pod is master
1. store all known sentinel ip addresses
1. sentinel process shuts down
1. if the pod was master
    1. wait for redis to shutdown
    1. wait for remaining sentinels to elect a new master
1. cycle through all sentinel ip addresses and call reset on them
    * so they forget about the container currently shutting down
1. container closes

In Redis Container:

1. wait for sentinel to shutdown
1. shut down the redis process
1. container closes

preStop only allows for handling what a container does before the process shuts down. I don't think that will work in this case since we need to add functionality to post process shutdown. I agree that it would be nice to find a way around using traps though.

We also changed the liveness probe back to only pinging the server again for both the sentinel and redis containers. Checking for NOQUORUM in the sentinel container caused the container to go down while other pods were in the middle of a rolling deploy. So pod A is taken down for the deployment, pod B checks if all sentinels are good and sees that pod A isn't, then pod B fails liveness and is restarted. We observed frequent cases of a healthy pod going down because another pod was being restarted as part of a rolling deployment. This sometimes also cascaded into the entire cluster going down.

Checking for the number of slaves in the redis container has the potential issue where if somebody sets min-slaves-to-write to something greater than 1, then it seems like no redis container could come up on initial launch since it'd be waiting for a number greater than 1 and it could never get that.

It would be nice if the liveness and readiness were a little better than a ping but it seems like they should only hit the pod in question and not reach out to the cluster for additional info.

ssalaues replied:
@jasquat @Globegitter I'm having trouble reproducing both your results as far as the quorum check. Also I'm not understanding what exactly is causing these issues outside having an unhealthy quorum or like you mentioned having `min-slaves-to-write` greater than 1 for a 3 replica deployment.

In your example, pod A being taken down for the rolling deploy would mean that pod B and C should both pass the NOQUORUM check. I just tested this manually by killing one pod and running the check via CLI on both pod B and pod C:
```
$ redis-cli -p 26379 sentinel ckquorum mymaster
OK 2 usable Sentinels. Quorum and failover authorization can be reached
```
Is it possible you are unknowingly testing this on an unhealthy "split" deployment?

ssalaues replied:
So upon further testing, I found that from upgrading from 3.0.3 in an unhealthy state (split or otherwise) to 3.0.4 caused a scenario that exactly matches what you both are reporting. At least from my "unhealthy" test environments, I found that the init would hang for maybe 1-2 minutes on the below line because it is waiting on a response of the reported master that no longer exists.
https://github.com/helm/charts/blob/master/stable/redis-ha/templates/redis-ha-configmap.yaml#L42

I tried the same scenario with a simple `timeout -t 3` in front of the redis-cli command and the rolling upgrade was able to successfully complete.

jasquat replied:
@ssalaues So for our testing we are using https://github.com/burnettk/redis-rb-cluster and using 3 bash sessions:
Session 1 (runs the actual test - deploys the pods and then performs a rolling restart on the redis cluster):
```
REDIS_HA_HELM_CHART_DIR=[path_to_github_charts]/charts/stable/redis-ha ./test_redis_ha [namespace]
```
Session 2 (command from session 1 will eventually create this pod):
```
kubect logs -f -n [namespace] deploy/redis-acceptance-test
```
Session 3 (can be started at anytime and prints out some status info using kubectl commands):
```
./redis_cluster_status [namespace]
```

From the cluster status command in the third session, you can watch the cluster go in and out of bad states as the cluster performs a rolling deployment. It is however inconsistent and doesn't always get into a bad state. Run it a couple of times though and you can see some badness.

The issue with the liveness probe is that a liveness probe should check the status of just the container that it represents. Currently the liveness probe is checking what the container thinks the cluster status is. So during a rolling deploy, the cluster status may periodically get into bad states and this may cause a perfectly healthy container to go down. Also, it doesn't make sense to me why the sentinel container should be restarted if there already aren't enough sentinels which is what checking the sentinel quorum on a sentinel's liveness does.

ssalaues replied:
Yeah I can see this from my end. I can revisit the quorum check at a later time likely as a "readiness" probe but will require further testing. For now I think we can push forward with #10521 to handle the probe issue while this is in review.

I'm reading/looking into your other comments as well.

ssalaues replied:
> preStop only allows for handling what a container does before the process shuts down. I don't think that will work in this case since we need to add functionality to post process shutdown.

The only post process shutdown functionality I can see is waiting for the new master and reset the sentinels which I'm not entirely convinced we need to do. Unless I'm missing some other reasoning, resetting the sentinels isn't ideal since you also have to do a check to ensure that the same master isn't elected. The same outcome can be done by doing a `redis-cli -p 26379 sentinel failover mymaster` which forces the quorum to select a new master. Once this is done, the same script can wait for the new master to be elected all part of the preStop.

I tested a simple preStop script that does all this against the acceptance test and the outage was never more than 3-4 seconds which seemed mostly to be the delay in the client receiving the update.

jasquat replied:
Yeah we tried using the sentinel failover command before but we did not have good results from it which is why we started resetting the sentinels instead. I can give this a try next week and see if I can see the same results using failover in a preStop script.

ssalaues replied:
It's possible you were having similar issues as https://github.com/helm/charts/issues/10635 where the sentinels would artificially grow causing a failover command to not work consistently. I remember I was originally testing the `sentinel reset` as a possible solution for that and was working well but ended up persisting the sentinel id instead to avoid the need for a reset.

Also if it helps clarify my comment, here's the preStop script I added to my sentinel containers and tested with. Likely over simplifying the shutdown (and lots of duplicate code snippets from the init.sh) but it's the basic idea of what I was getting at.
```
shutdown.sh: |    
    set -e
    HOSTNAME="$(hostname)"
    INDEX="${HOSTNAME##*-}"
    MASTER="$(redis-cli -h {{ template "redis-ha.fullname" . }} -p {{ .Values.sentinel.port }} sentinel get-master-addr-by-name {{ .Values.redis.masterGroupName }} | grep -E '[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}')"
    MASTER_GROUP="{{ .Values.redis.masterGroupName }}"
    REDIS_PORT={{ .Values.redis.port }}
    SENTINEL_PORT={{ .Values.sentinel.port }}
    SERVICE={{ template "redis-ha.fullname" . }}
    echo "Shutdown process initiated"
    if getent hosts $MASTER | grep -q "announce-$INDEX" ; then
        redis-cli -h "$SERVICE" -p "$SENTINEL_PORT" sentinel failover "$MASTER_GROUP"
        sleep 20
        while [ $MASTER = "$(redis-cli -h {{ template "redis-ha.fullname" . }} -p {{ .Values.sentinel.port }} sentinel get-master-addr-by-name {{ .Values.redis.masterGroupName }} | grep -E '[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}')" ]; do
          sleep 1
        done
    fi
    redis-cli -h 127.0.0.1 -p "$SENTINEL_PORT" sentinel shutdown
```

ssalaues reviewed (commented):
A much needed direction to take the chart in but can simplify quite a bit using container hooks.

Commit:
fix: reference min slaves in healthcheck

Signed-off-by: Salim <salim.salaues@scality.com>

jasquat committed:
Merge remote-tracking branch 'origin/master' into cluster-aware-shutdown

jasquat committed:
merged some changes from a branch later merged into master. w/ burnettk

jasquat committed:
merged in master. w/ burnettk

jasquat committed:
[stable/redis-ha] removed debug code and use ping for liveness since
checking cluster for health causes various containers to go down during
a rollout.

Signed-off-by: Jason Lantz <jasquat@gmail.com>

jasquat committed:
[stable/redis-ha] updated version number

Signed-off-by: Jason Lantz <jasquat@gmail.com>
